rfcs/0037-x86_32.md

```diff
@@ -0,0 +1,57 @@
+---
+feature: x86_32
+start-date: 2018-11-21
+author: Daniel Goertzen (@goertzenator)
+co-authors: (find a buddy later to help our with the RFC)
+related-issues: 52634
+---
+
+# Summary
+[summary]: #summary
+
+Expand x86 32 bit support beyond i686 to include i586, i486, and i386.  This biggest part of this change is to disambiguate usages of `i686` which is currently overloaded to mean both "32 bit x86" and "exactly the i686 architecture".  The term `x86_32` is proposed here to replace many current usages of `i686`.
+
+
+# Motivation
+[motivation]: #motivation
+
+- To support embedded x86 SoCs such as the i586 [Vortex86](https://en.wikipedia.org/wiki/Vortex86).  I personally would like to use nix on such a CPU.
+- To provide the framework for other contributers to expand support for these architectures.
+
+# Detailed design
+[design]: #detailed-design
+
+The primary part that requires discussion is disambiguation of `i686`.  After that is resolved the addition of `i[345]86` support is less controversial.
+
+Introduce the predicate `isx86_32` which is symmetric to the already existing `isx86_64`.  `isx86_32` will mean "32 bit x86" and the existing `isi686` will be reinterpreted as "exactly i686 architecture".
+
+For i686 this change is effectively a low-risk no-op because `isx86_32` and `isi686` will produce the same result.  Only when the CPU is changed to `i[345]86` does the distinction begin to matter.
+
+Once the above change is agreed upon then the addition of `i[345]86` CPUs becomes straight forward.  Predicates `isi586`, `isi485`, and `isi386` will be added.  Usages of `isi686` in various packages will be flipped to `isx86_32` as needed to make them work on these CPUs.
+
+Pragmatically, I will be looking to enable i586 support on several common packages including linux kernel, libc, and openssh.  I'm not interested in i486 and i386, but it is trivial to add these CPU definitions alongside i586 should someone in future want to take a crack at it.
+
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- i586 and predecessors are already uncommon and use continues to diminish.
```

Comment:
The usual issue we have with things like this (minority architectures/platforms) is that someone strongly motivated will propose something, do a lot of cross-cutting work to support it across all of nixpkgs, then lose steam and then the maintenance burden of all the cross-cutting changes falls upon the community.

So I'd phrase the drawbacks in terms of maintenance burden and degree to which this support can be confined to a small region of nixpkgs. The more "compact" such support is, the easier it is for me to believe it's practical to put into nixpkgs. If it's straggly (all sorts of random switches in all sorts of random builds) it becomes a much harder sell to me. I've already been on the fence about the cost/benefit ratio of our i686 support, not because I don't want more people to benefit from Nix, but because being the only maintainer of a niche platform within a niche package manager is not fun and leads to burnout when everyone's patches are constantly breaking your stuff. The biggest example we have of this is macOS, which I think has achieved "critical mass" (to the point of being sustainable), often had me questioning it along the way. Other Linux users probably still do question the value of the occasional `.isDarwin` switch floating around their packages that they can't test and can't engage meaningfully on.